src/abstract/BaseHookUpgradeable.sol

```diff
@@ -0,0 +1,269 @@
+        address,
+        PoolKey calldata,
+        IPoolManager.ModifyLiquidityParams calldata,
+        bytes calldata
+    ) internal virtual returns (bytes4) {
+        revert HookNotImplemented();
+    }
+
+    /// @inheritdoc IHooks
+    function afterAddLiquidity(
+        address sender,
+        PoolKey calldata key,
+        IPoolManager.ModifyLiquidityParams calldata params,
+        BalanceDelta delta,
+        BalanceDelta feesAccrued,
+        bytes calldata hookData
+    ) external onlyPoolManager returns (bytes4, BalanceDelta) {
+        return _afterAddLiquidity(sender, key, params, delta, feesAccrued, hookData);
+    }
+
+    function _afterAddLiquidity(
+        address,
+        PoolKey calldata,
+        IPoolManager.ModifyLiquidityParams calldata,
+        BalanceDelta,
+        BalanceDelta,
+        bytes calldata
+    ) internal virtual returns (bytes4, BalanceDelta) {
+        revert HookNotImplemented();
+    }
+
+    /// @inheritdoc IHooks
+    function afterRemoveLiquidity(
+        address sender,
+        PoolKey calldata key,
+        IPoolManager.ModifyLiquidityParams calldata params,
+        BalanceDelta delta,
+        BalanceDelta feesAccrued,
+        bytes calldata hookData
+    ) external onlyPoolManager returns (bytes4, BalanceDelta) {
+        return _afterRemoveLiquidity(sender, key, params, delta, feesAccrued, hookData);
+    }
+
+    function _afterRemoveLiquidity(
+        address,
+        PoolKey calldata,
+        IPoolManager.ModifyLiquidityParams calldata,
+        BalanceDelta,
+        BalanceDelta,
+        bytes calldata
+    ) internal virtual returns (bytes4, BalanceDelta) {
+        revert HookNotImplemented();
+    }
+
+    /// @inheritdoc IHooks
+    function beforeSwap(
+        address sender,
+        PoolKey calldata key,
+        IPoolManager.SwapParams calldata params,
+        bytes calldata hookData
+    ) external onlyPoolManager returns (bytes4, BeforeSwapDelta, uint24) {
+        return _beforeSwap(sender, key, params, hookData);
+    }
+
+    function _beforeSwap(
+        address,
+        PoolKey calldata,
+        IPoolManager.SwapParams calldata,
+        bytes calldata
+    ) internal virtual returns (bytes4, BeforeSwapDelta, uint24) {
+        revert HookNotImplemented();
+    }
+
+    /// @inheritdoc IHooks
+    function afterSwap(
+        address sender,
+        PoolKey calldata key,
+        IPoolManager.SwapParams calldata params,
+        BalanceDelta delta,
+        bytes calldata hookData
+    ) external onlyPoolManager returns (bytes4, int128) {
+        return _afterSwap(sender, key, params, delta, hookData);
+    }
+
+    function _afterSwap(
+        address,
+        PoolKey calldata,
+        IPoolManager.SwapParams calldata,
+        BalanceDelta,
+        bytes calldata
+    ) internal virtual returns (bytes4, int128) {
+        revert HookNotImplemented();
+    }
+
+    /// @inheritdoc IHooks
+    function beforeDonate(
```

Comment:
hook address will be mined to include only certain bits as enabled. So I don't think that upgrade can actually enable functions which were not used before. Worth checking it out

Reply:
Yes, that's one limitation of the hooks design. 
Not sure which features we would add to these hooks but for example for the Allowlist one, we would be able to add the swap cap via an upgrade since `beforeSwap` and `afterSwap` are activated at deployment.